api/proto/config/storage.proto

```diff
@@ -242,3 +242,27 @@ message DisksConfig {
   DisksArrayType array_type = 2;
   repeated DisksConfig children = 3;
 }
+
+// ZfsOpType is the commands to execute in ZFS
+enum ZfsOpType {
+  ZFS_OP_TYPE_UNSPECIFIED = 0;
+  ZFS_OP_TYPE_SNAPSHOT_CREATE = 1;
+  ZFS_OP_TYPE_SNAPSHOT_DELETE = 2;
+  ZFS_OP_TYPE_SNAPSHOT_RENAME = 3;
+  ZFS_OP_TYPE_SNAPSHOT_ROLLBACK = 4;
+}
+
+// ZfsResponseOpType is the response from ZFS after run command
+enum ZfsResponseOpType {
+  ZFS_RESPONSE_OP_TYPE_UNSPECIFIED = 0;
+  ZFS_RESPONSE_OP_TYPE_SUCCESS = 1;
+  ZFS_RESPONSE_OP_TYPE_ERROR = 2;
+}
+
+// ZfsSnapshotCommand executes a command to work with snapshots in ZFS
+message ZfsSnapshotCommand {
```

Comment:
I don't see how this will be used but in general we try very hard to make the EVE API be declarative e.g., in the API we specify the set of objects (e.g., app instances) which should exist on the device. This means that there is no delete; the removal of an object from the set means that it should no longer exist which will result in EVE-OS deleting it.

There are a small number of places where we have imperative commands such as rebooting EVE-OS or restarting an app instance, but those must be expressed as declarative statements which we do by using a counter together with the command.

I suspect that for snapshots we can take a declarative approach and instead of having an imperative rename command we can have support for annotating a snapshot with a description (or that could potentially be done entirely in the controller).

But do we have a document explaining the use cases and end-to-end workflows for snapshots?  It would make sense to have that as the basis for the API needs.

Also, do we know whether we need to interact with the guest VM to pause modifications against the zvol to make sure that the snapshot can be used in a roolback?

Reply:
@eriknordmark we do not need to interact with VM since ZFS pauses the I/O on that zvol before taking snapshot.